### PR TITLE
Fix libdeflate linking for vendored htslib

### PR DESCRIPTION
**Pull request by jdblischak**
**tl;dr** The vendored htslib isn't linked to libdeflate

**Note:** This PR is intended to initially fail to demonstrate the problem. I already know how to fix it, and will push those commit(s) next

## History of vendored htslib

We recently replaced htslib from bioconda with a vendored htslib build (PR #64; commit a5c01af; version 0.20.2, build 1). Furthermore, since conda-forge only pins a single libdeflate version, we expanded to build for multiple libdeflate versions (PRs #67  and #70)

## Demonstration of libdeflate linking

From the [htslib installation instructions](https://github.com/samtools/htslib/blob/6652c86ab34aed5aaf3e1211ba75cea982ad269c/INSTALL#L175), the default behavior is to search for libdeflate and link to it if it is available. If it can't find it, there is no error. Since cmake suppresses the build logs of external dependencies like htslib, it's not possible to determine from the build logs if the linking occurred or not.

The [bioconda build script](https://github.com/bioconda/bioconda-recipes/blob/6cb443874a8e7c089e76aef7a31eb7eb94416b4f/recipes/htslib/build.sh#L3) explicitly sets `--with-libdeflate`, which causes the htslib build to fail if no libdeflate can be found. I added this to our htslib cmake patch, which then causes the build to fail.

Using `ldd`:

```sh
# htslib from bioconda
mamba create --yes -n test-htslib-bioconda -c bioconda htslib
mamba activate test-htslib-bioconda
ls $CONDA_PREFIX/lib/libhts.*
## $CONDA_PREFIX/lib/libhts.a
## $CONDA_PREFIX/lib/libhts.so
## $CONDA_PREFIX/lib/libhts.so.1.9
## $CONDA_PREFIX/lib/libhts.so.2
ldd $CONDA_PREFIX/lib/libhts.so* | grep libdeflate
##         libdeflate.so => $CONDA_PREFIX/lib/./libdeflate.so (0x00007f3643efa000)
##         libdeflate.so => $CONDA_PREFIX/lib/./libdeflate.so (0x00007f61a4fbc000)
##         libdeflate.so => $CONDA_PREFIX/lib/./libdeflate.so (0x00007fb09c062000)
mamba deactivate

# htslib vendored in libtiledbvcf
mamba create --yes -n test-htslib-tiledbvcf -c conda-forge -c tiledb 'libtiledbvcf=0.22.0=*2'
mamba activate test-htslib-tiledbvcf
mamba list libtiledbvcf
## # Name                    Version                   Build  Channel
## libtiledbvcf              0.22.0               h3156680_2    tiledb
ls $CONDA_PREFIX/lib/*tile*
## $CONDA_PREFIX/lib/libtiledb.so
## $CONDA_PREFIX/lib/libtiledb.so.2.14
## $CONDA_PREFIX/lib/libtiledbvcf.so
ls $CONDA_PREFIX/lib/*hts*
## $CONDA_PREFIX/lib/libhts.so.1.15.1
ldd $CONDA_PREFIX/lib/libtiledbvcf.so | grep libhts
##         libhts.so.1.15.1 => $CONDA_PREFIX/lib/./libhts.so.1.15.1 (0x00007fcb152e2000)
ldd $CONDA_PREFIX/lib/libhts.so.1.15.1 | grep libdeflate
mamba deactivate
```

## Cause

I have always been confused why this recipe includes top-level requirements which are largely redundant to the requirements for the subpackage libtiledbvcf. I previously deleted them, but the builds would always fail. I finally figured it out from this comment https://github.com/conda-forge/conda-forge.github.io/issues/746#issuecomment-937045471 and inspecting the build logs. Essentially, the top-level requirements are installed prior to running the default `build.sh`.

But libdeflate was only added to the requirements in the subpackage section, which simply runs `make install-libtiledbvcf` (in `install-libtiledbvcf.sh`). htslib is built by the call to cmake in `build.sh`, when no libdeflate is available.

## Solution

To both simplify this recipe and fix the linking to libdeflate, I propose combining the code in `build.sh` and `install-libtiledbvcf.sh` into a single file (eg `build-libtiledbvcf.sh`) and removing the top-level requirements section. This will result in libdeflate being available during the htslib build, and it will ease future maintenance (only one requirements section to update), and it will reduce CI time (because running `build.sh` separately requires solving and installing an entirely separate conda env).

### CODE REVIEW

**Comment by jdblischak** [comment]
I realized that unfortunately my current solution doesn't completely solve the problem. The conda-installed libdeflate is now available when htslib is compiled, however it still can't find libdeflate for some reason. I'm currently tinkering with `CFLAGS` and `LDFLAGS`, as suggested in the [installation instructions](https://github.com/samtools/htslib/blob/6652c86ab34aed5aaf3e1211ba75cea982ad269c/INSTALL#L196):

> If dependencies have been installed in non-standard locations (i.e. not on
> the normal include and library search paths) then the CPPFLAGS and LDFLAGS
> environment variables can be used to set the options needed to find them.

I assume I need to somehow adjust the search paths by altering the env vars or flags passed to cmake or to directly patch the env vars or flags in [`FindHTSlib.cmake`](https://github.com/TileDB-Inc/TileDB-VCF/blob/master/libtiledbvcf/cmake/Modules/FindHTSlib.cmake). I'd appreciate any suggestions

**Comment by jdblischak** [comment]
This PR was too ambitious. Even after installing libdeflate in the conda env that runs `cmake`, it still isn't properly linked. I tried patching the upstream `FindHTSlib.cmake` to properly link libdeflate, but trying to troubleshoot the cmake instructions via a patched conda build running on Azure was too cumbersome. I tried setting the same flags as [the bioconda::pysam build script](https://github.com/bioconda/bioconda-recipes/blob/7aeabde7832b90b020c7d08dfd220f5370efa2f7/recipes/pysam/build.sh#L6) uses, but it failed without providing a useful error message.

```cmake
./configure --disable-libcurl --disable-gcs --disable-aws --prefix=${EP_INSTALL_PREFIX} LDFLAGS="${EXTRA_LDFLAGS},-L${EP_INSTALL_PREFIX}/lib" CFLAGS="${CFLAGS},-I${EP_INSTALL_PREFIX}/include -DHAVE_LIBDEFLATE" CPPFLAGS="${CPPFLAGS},-I${EP_INSTALL_PREFIX}/include -DHAVE_LIBDEFLATE"
```

For now I am planning workaround this problem using conda recipes. This linking difficulty may not even be a problem for users that build libtiledbvcf from source using the cmake superbuild, since it's possible that it is easier for htslib to find and link libdeflate when everything is installed in standard system directories (as opposed to a non-standard conda env directory).
